lib/krane/kubernetes_resource.rb

```diff
@@ -409,17 +414,61 @@ def self.go_template_for(kind, name)
       def self.extract_all_from_go_template_blob(blob)
         blob.split(EVENT_SEPARATOR).map do |event_blob|
           pieces = event_blob.split(FIELD_SEPARATOR, FIELDS.length)
+          count = extract_event_count(pieces)
+          timestamp = extract_event_timestamp(pieces)
+
           new(
             subject_kind: pieces[FIELDS.index(".involvedObject.kind")],
             subject_name: pieces[FIELDS.index(".involvedObject.name")],
-            count: pieces[FIELDS.index(".count")],
-            last_timestamp: pieces[FIELDS.index(".lastTimestamp")],
+            count: count,
+            last_timestamp: timestamp,
             reason: pieces[FIELDS.index(".reason")],
             message: pieces[FIELDS.index(".message")]
           )
         end
       end
 
+      def self.extract_event_count(pieces)
+        series = pieces[FIELDS.index(".series")]
+        count = pieces[FIELDS.index(".count")]
+        deprecated_count = pieces[FIELDS.index(".deprecatedCount")]
+
+        # Find the right event count according to Kubernetes API and kubectl version
+        if count.present? && count != FIELD_EMPTY_VALUE
+          count # This is the default field, so let's try to use it first
+        elsif series.present? && series != FIELD_EMPTY_VALUE
+          # kubectl 1.16 uses Events/v1, which has the .series/.count field
+          count_regex = /count:(?<value>\S+?(?=\s))/
+          count_regex.match(series)['value']
+        elsif deprecated_count.present? && deprecated_count != FIELD_EMPTY_VALUE
+          # kubectl < 1.16 uses events.k8s.io/v1beta1, which has .deprecatedCount
+          deprecated_count
+        else
+          "1" # Fallback to 1 when all count fields are null
```

Comment:
The API reference defines count as an integer, how come we're using a String?

Reply:
good point, that reference https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.16/#event-v1-core does show it's an integer, 😕

Reply:
I'm using `"1"` to be consistent which is returned by `pieces[FIELDS.index(".count")]` or `pieces[FIELDS.index(".deprecatedCount")]`:

```ruby
kubectl get events --output=go-template={{range .items}}{{if and (eq .involvedObject.kind "Pod") (eq .involvedObject.name "pvc") (ne .reason "Started") (ne .reason "Created") (ne .reason "SuccessfulCreate") (ne .reason "Scheduled") (ne .reason "Pulling") (ne .reason "Pulled")}}{{.involvedObject.kind}}{{print "ENDFIELD--BEGINFIELD"}}{{.involvedObject.name}}{{print "ENDFIELD--BEGINFIELD"}}{{.count}}{{print "ENDFIELD--BEGINFIELD"}}{{.lastTimestamp}}{{print "ENDFIELD--BEGINFIELD"}}{{.reason}}{{print "ENDFIELD--BEGINFIELD"}}{{.message}}{{print "ENDFIELD--BEGINFIELD"}}{{.eventTime}}{{print "ENDFIELD--BEGINFIELD"}}{{.deprecatedCount}}{{print "ENDFIELD--BEGINFIELD"}}{{.deprecatedLastTimestamp}}{{print "ENDFIELD--BEGINFIELD"}}{{.series}}{{print "ENDEVENT--BEGINEVENT"}}{{end}}{{end}} --namespace=k8sdeploy-test-pvc-no-pv-4e96cca0a0d64155 --context=minikube --request-timeout=15

From: /Users/douglas/src/github.com/Shopify/kubernetes-deploy/lib/krane/kubernetes_resource.rb @ line 422 Krane::KubernetesResource::Event.extract_all_from_go_template_blob:

    414: def self.extract_all_from_go_template_blob(blob)
    415:   blob.split(EVENT_SEPARATOR).map do |event_blob|
    416:     pieces = event_blob.split(FIELD_SEPARATOR, FIELDS.length)
    417:     count = extract_event_count(pieces)
    418:     timestamp = extract_event_timestamp(pieces)
    419:
    420:     binding.pry
    421:
 => 422:     new(
    423:       subject_kind: pieces[FIELDS.index(".involvedObject.kind")],
    424:       subject_name: pieces[FIELDS.index(".involvedObject.name")],
    425:       count: count,
    426:       last_timestamp: timestamp,
    427:       reason: pieces[FIELDS.index(".reason")],
    428:       message: pieces[FIELDS.index(".message")]
    429:     )
    430:   end
    431: end
[1] pry(Krane::KubernetesResource::Event)> pieces[FIELDS.index(".count")]
=> "4"
```

After we get the correct `count` value, it is passed to `Event`initializer: 
 https://github.com/Shopify/kubernetes-deploy/blob/fix_krane_k8s_api_1.16/lib/krane/kubernetes_resource.rb#L478 which does `@count = count.to_i`, so we're still respecting k8s API.

Reply:
I'm using `"1"` to be consistent which is returned by `pieces[FIELDS.index(".count")]` or `pieces[FIELDS.index(".deprecatedCount")]`:

```ruby
kubectl get events --output=go-template={{range .items}}{{if and (eq .involvedObject.kind "Pod") (eq .involvedObject.name "pvc") (ne .reason "Started") (ne .reason "Created") (ne .reason "SuccessfulCreate") (ne .reason "Scheduled") (ne .reason "Pulling") (ne .reason "Pulled")}}{{.involvedObject.kind}}{{print "ENDFIELD--BEGINFIELD"}}{{.involvedObject.name}}{{print "ENDFIELD--BEGINFIELD"}}{{.count}}{{print "ENDFIELD--BEGINFIELD"}}{{.lastTimestamp}}{{print "ENDFIELD--BEGINFIELD"}}{{.reason}}{{print "ENDFIELD--BEGINFIELD"}}{{.message}}{{print "ENDFIELD--BEGINFIELD"}}{{.eventTime}}{{print "ENDFIELD--BEGINFIELD"}}{{.deprecatedCount}}{{print "ENDFIELD--BEGINFIELD"}}{{.deprecatedLastTimestamp}}{{print "ENDFIELD--BEGINFIELD"}}{{.series}}{{print "ENDEVENT--BEGINEVENT"}}{{end}}{{end}} --namespace=k8sdeploy-test-pvc-no-pv-4e96cca0a0d64155 --context=minikube --request-timeout=15

From: /Users/douglas/src/github.com/Shopify/kubernetes-deploy/lib/krane/kubernetes_resource.rb @ line 422 Krane::KubernetesResource::Event.extract_all_from_go_template_blob:

    414: def self.extract_all_from_go_template_blob(blob)
    415:   blob.split(EVENT_SEPARATOR).map do |event_blob|
    416:     pieces = event_blob.split(FIELD_SEPARATOR, FIELDS.length)
    417:     count = extract_event_count(pieces)
    418:     timestamp = extract_event_timestamp(pieces)
    419:
    420:     binding.pry
    421:
 => 422:     new(
    423:       subject_kind: pieces[FIELDS.index(".involvedObject.kind")],
    424:       subject_name: pieces[FIELDS.index(".involvedObject.name")],
    425:       count: count,
    426:       last_timestamp: timestamp,
    427:       reason: pieces[FIELDS.index(".reason")],
    428:       message: pieces[FIELDS.index(".message")]
    429:     )
    430:   end
    431: end
[1] pry(Krane::KubernetesResource::Event)> pieces[FIELDS.index(".count")]
=> "4"
```

After we get the correct `count` value, it is passed to `Event`initializer: 
 https://github.com/Shopify/kubernetes-deploy/blob/fix_krane_k8s_api_1.16/lib/krane/kubernetes_resource.rb#L478 which does `@count = count.to_i`, so we're still respecting k8s API.

Reply:
Thanks for the prove, good explanation @douglas